swift/apple/FirezoneKit/Sources/FirezoneKit/Views/MenuBar.swift

```diff
@@ -553,6 +632,35 @@ public final class MenuBar: NSObject, ObservableObject {
     }
   }
 
+  @objc private func addFavoriteTapped(_ sender: NSMenuItem) {
+    guard let id = sender.representedObject as? String else { return }
+    setFavorited(id: id, favorited: true)
+  }
+
+  @objc private func removeFavoriteTapped(_ sender: NSMenuItem) {
+    guard let id = sender.representedObject as? String else { return }
+    setFavorited(id: id, favorited: false)
+  }
+
+  private func setFavorited(id: String, favorited: Bool) {
+    if favorited {
+      favorites.add(id)
+    } else {
+      favorites.remove(id)
+    }
+    favoritesChanged()
+  }
+
+  func favoritesChanged() {
+    let newResources = resources.map { res in
+      ViewResource(base: res.base, isFavorite: favorites.contains(res.base.id))
+    }
+    // When the user clicks to add or remove a favorite, the menu will close anyway, so just recreate the whole menu.
+    // This avoids complex logic when changing in and out of the "nothing is favorited" special case
+    self.populateResourceMenus([])
+    self.populateResourceMenus(newResources)
```

Comment:
do we need to do this? from what I remember implementing the disable resources when the menu opens again won't it be repopulated?

Reply:
This is a hack because the diffing doesn't work correctly or something. Looking into it :/

Reply:
It might only apply to the first Resource in the list but basically it's like this:
- For backwards compatibility we hide the "Other Resources" menu when nothing is favorited
- That means when you switch between 1 favorite and 0 favorites, one item is always staying in the favorites menu even though its favorite / non-favorite status is toggling
- With `ViewResource` gone, that means it no longer shows up in the diff
- Even when `ViewResource` existed, the diff acted weird with respect to favorites
- Since that one item stays in the same menu, it never gets refreshed

Reply:
I spent a couple hours trying to get rid of this and not dismiss the menu but it's looking like a dead end, I just don't know `NSMenu` well enough to implement it in a simple way, and rewriting the whole `MenuBar.swift` would take too long

Reply:
All good, all of the other button actions in the menu dismiss it so it's probably ok.